### PR TITLE
Introduce libs/rays and implement AO baking.

### PR DESCRIPTION
**Pull request by prideout**
The AO baking procedure consists of the following steps:

1. Flatten the glTF hierarchy.
2. Generate a single 2D parameterization for the entire scene.
3. Embree Pass 1: Create G-Buffer using the above UVs as vert positions.
4. Embree Pass 2: Cast rays from the positions embedded in the G-Buffer.

The `gltf_baker` tool is not ready for general use but already produces reasonable results for certain well-formed models.

Screen Shot 2019-05-14 at 1 54 44 PM

### CODE REVIEW

**Comment by prideout** [comment]
I'll look into the Android build.

File: libs/rays/src/PathTracer.cpp [hunk 1]
```diff
@@ -0,0 +1,480 @@
...
+
+    if (mConfig.uvCamera) {
+
+        context->scene = rtcNewScene(device);
+        populate2DScene();
+
+        context->gbuffer = image::LinearImage(width, height, 6);
+
+        // First render XYZ positions into the G-Buffer.
+        spawnTileJobs(context, [populate3DScene](EmbreeContext* context, PixelRectangle rect) {
```
Comment (bejado) [inline libs/rays/src/PathTracer.cpp]: Do you need to capture `populate3DScene` here?

File: libs/rays/src/PathTracer.cpp [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+        }
+        rtcCommitScene(context->scene);
+    };
+
+    if (mConfig.uvCamera) {
+
+        context->scene = rtcNewScene(device);
+        populate2DScene();
+
+        context->gbuffer = image::LinearImage(width, height, 6);
```
Comment (bejado) [inline libs/rays/src/PathTracer.cpp]: Maybe a comment explaining the layout of the `gbuffer`? Is it 3 position floats followed by 3 normal floats?

File: libs/rays/src/PathTracer.cpp [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+            rtcInitIntersectContext(&intersector);
+
+            RTCRay ray = generateOrthoCameraRay(row, col);
+            RTCRayHit rayhit { .ray = ray };
+            intersector.flags = RTC_INTERSECT_CONTEXT_FLAG_COHERENT;
+            rtcIntersect1(embreeScene, &intersector, &rayhit);
+
+            if (rayhit.ray.tfar != inf) {
+                float* position = gbuffer.getPixelRef(col, row);
+                float* normal = position + 3;
```
Comment (bejado) [inline libs/rays/src/PathTracer.cpp]: Might be good to define the `3` here as a constant, which will help with readability and if the layout of the gbuffer ever changes.

Comment (prideout) [reply libs/rays/src/PathTracer.cpp]: Done.

File: libs/rays/src/PathTracer.cpp (hunk 2 above)
Comment (prideout) [reply libs/rays/src/PathTracer.cpp]: Done.

File: libs/rays/src/PathTracer.cpp (hunk 1 above)
Comment (prideout) [reply libs/rays/src/PathTracer.cpp]: Fixed.

**Pull request merged**